File: hammer-gui/src/main/java/dev/vini2003/hammer/gui/api/client/event/InGameHudEvents.java
package dev.vini2003.hammer.gui.api.client.event;

import dev.vini2003.hammer.gui.api.common.widget.WidgetCollection;
import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.client.gui.hud.InGameHud;
import net.minecraft.client.render.VertexConsumerProvider;
import net.minecraft.client.util.math.MatrixStack;

public interface InGameHudEvents {
	Event<Init> INIT = EventFactory.createArrayBacked(Init.class, (events) -> (hud, collection) -> {
		for (var event : events) {
			event.onInit(hud, collection);
		}
	});
	
	Event<Render> RENDER = EventFactory.createArrayBacked(Render.class, (events) -> (matrices, provider, hud, collection) -> {
		for (var event : events) {
			event.onRender(matrices, provider, hud, collection);
		}
	});
	
	@FunctionalInterface
	interface Init {
		void onInit(InGameHud hud, WidgetCollection collection);
	}
	
	@FunctionalInterface
	interface Render {
		void onRender(MatrixStack matrices, VertexConsumerProvider.Immediate provider, InGameHud hud, WidgetCollection collection);
	}
}

File: hammer-gui/src/main/kotlin/dev/vini2003/hammer/gui/api/client/event/InGameHudEvents.kt
package dev.vini2003.hammer.gui.api.client.event

import dev.vini2003.hammer.gui.api.common.widget.BaseWidgetCollection
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.client.gui.hud.InGameHud
import net.minecraft.client.render.VertexConsumerProvider
import net.minecraft.client.util.math.MatrixStack

interface InGameHudEvents {
	companion object {
		@JvmField
		val INIT = EventFactory.createArrayBacked(Init::class.java) { events ->
			object : Init {
				override fun onInit(hud: InGameHud, collection: BaseWidgetCollection) {
					events.forEach { event ->
						event.onInit(hud, collection)
					}
				}
			}
		}
		
		@JvmField
		val RENDER = EventFactory.createArrayBacked(Render::class.java) { events ->
			object : Render {
				override fun onRender(matrices: MatrixStack, provider: VertexConsumerProvider.Immediate, hud: InGameHud, collection: BaseWidgetCollection) {
					events.forEach { event ->
						event.onRender(matrices, provider, hud, collection)
					}
				}
			}
		}
	}
	
	@FunctionalInterface
	fun interface Init {
		fun onInit(hud: InGameHud, collection: BaseWidgetCollection)
	}
	
	@FunctionalInterface
	fun interface Render {
		fun onRender(matrices: MatrixStack, provider: VertexConsumerProvider.Immediate, hud: InGameHud, collection: BaseWidgetCollection)
	}
}